### PR TITLE
yelp api v3 (yelp-fusion)

### LINKED ISSUE #4: New API Authentication Method
Does somebody know how to update this to use the new Yelp API authentication method?

### LINKED ISSUE #6: Update Provider to work with new Yelp Fusion API
This provider uses the [yelp npm package](https://www.npmjs.com/package/yelp) ([here](https://github.com/koopjs/koop-provider-yelp/blob/bb26d2735b95183441fbe7087004205268504ae5/package.json#L20)), which says, ` Javascript client for Yelp's API (v2.0)`. [Yelp API v2 was discontinued on June 30, 2018](https://engineeringblog.yelp.com/2017/06/upcoming-deprecation-of-yelp-api-v2.html). 

That means **this provider is not currently working** - we need to update it to work with the new Yelp API ("[Yelp Fusion](https://www.yelp.com/fusion)"). The best Node package looks like [yelp-fusion](https://www.npmjs.com/package/yelp-fusion).

### PR DESCRIPTION
**Pull request by gavinr**
resolves #6
resolves #4

- uses yelp api v3 (yelp-fusion)
- includes unit tests
- updates structure to match modern koop provider structure

### CODE REVIEW

File: src/model.js
Comment (rgwozdz) [inline src/model.js]: You don't need to check `req.query` since Express will always have the `query` attribute on the request object.

File: src/model.js
Comment (rgwozdz) [inline src/model.js]: I don't understand why the count value is hard-coded and applied and returned here.  Are you certain that any type of request will always return this count?  It seems unlikely - what about the case where a `where` filter is passed that leads to no features returned?  e.g `?where=1=2&returnCountOnly` - that should return a count of `0`.

The FeatureServer dependency handles `returnCountOnly` and packages the response accordingly. Is there a reason not to let it do this job?

File: src/model.js
Comment (rgwozdz) [inline src/model.js]: No need to use `queries` variable here or `else`.  Just use `return`, i.e.,
```
if (options.geometry) {
  return [buildQuery(options, options.geometry)];
} 
```

File: src/model.js
Comment (rgwozdz) [inline src/model.js]: Why use async.each?  It looks like you `queries` is always a one element array.

File: src/model.js
Comment (rgwozdz) [inline src/model.js:67]: I think you should probably do this now instead of random timer.  Maybe use exponential backoff.

File: src/model.js
Comment (rgwozdz) [inline src/model.js]: Log should be removed.

**Comment by haoliangyu** [comment]
Wow, great PR! I am adding this to my review list.

File: src/model.js
Comment (rgwozdz) [inline src/model.js:137]: You can't count on the geometry parameter have max and mins. Geoservices API supports different data structures for geometry
```
geometry={xmin: -104, ymin: 35.6, xmax: -94.32, ymax: 41}
geometry=-104,35.6,-94.32,41
```

File: src/model.js
Comment (haoliangyu) [inline src/model.js]: Why is there a hard-coded number?

File: src/model.js
Comment (rgwozdz) [inline src/model.js:177]: Same as above.  Geometry might be an array of floats.

File: src/model.js
Comment (rgwozdz) [inline src/model.js:209]: If you are going to use `return` inside your `if` block (which is preferred), there is no reason to use `else`.

File: src/model.js
Comment (rgwozdz) [inline src/model.js]: No reason to use `else` because you have a return in your `if`.

File: src/model.js
Comment (haoliangyu) [inline src/model.js:67]: I think we can also implement a request queue with a fixed dispatch rate. But it is not necessary in this PR.

File: src/model.js
Comment (haoliangyu) [inline src/model.js]: Can you remove the console.log?

File: src/model.js
Comment (haoliangyu) [inline src/model.js:118]: Why `St. Louis, MO`?

**Review by haoliangyu — COMMENTED** [review]
Only some minor questions

File: src/model.js
Comment (rgwozdz) [inline src/model.js]: Please don't use slang like `biz` or `loc` - I know it's less keystrokes, but it take longer to read and over the lifetime of the repo quick readability probably more import. So use `business` and `location`.  Then use destructuring to get the properties your need.

```
const { 
  location,
  coordinates: { latitude, longitude },
  id,
  alias,
  // etc
} = business
```

Also If all you are using from the `query` parmeter is  `term`,  just pass in `term`.

**Comment by rgwozdz** [comment]
@gavinr - also if you are going to have the Yelp API do `geometry` and `where`  you should add `filtersApplied` object to the root of your GeoJSON object so that Koop does not try to do it a second time. Here's the spec:
```
filtersApplied: {
    geometry: Boolean, // true if a geometric filter has already been applied to the data
    where: Boolean, // true if a sql-like where filter has already been applied to the data
    offset: Boolean // true if the result offset has already been applied to the data,
    limit: Boolean // true if the result count has already been limited,
  }
```

File: src/model.js
Comment (haoliangyu) [reply src/model.js:67]: I disagree with the `exponential backoff`. All requests are valid and we still want them to finish as soon as possible (within limit).

Leaking bucket is a better solution.

Comment (rgwozdz) [reply src/model.js:67]: @haoliangyu good thinking, totally right

File: src/model.js
Comment (gavinr) [reply src/model.js]: This is [how the provider worked before](https://github.com/koopjs/koop-provider-yelp/blob/bb26d2735b95183441fbe7087004205268504ae5/yelp.js#L13-L15). I had assumed that was there for a specific purpose, but maybe it's not?

File: src/model.js
Comment (gavinr) [reply src/model.js]: done. thanks!

File: src/model.js
Comment (gavinr) [reply src/model.js:118]: Why not? :)

File: src/model.js
Comment (rgwozdz) [reply src/model.js]: I'd remove it for now then.

File: src/model.js
Comment (gavinr) [reply src/model.js]: Yeah, the multiple queries was a hold-over from the previous code, I forgot to remove this. thanks.

File: src/model.js
Comment (gavinr) [reply src/model.js]: similar to above. will remove.

File: src/model.js
Comment (gavinr) [reply src/model.js]: Removed. Thanks!

File: src/model.js
Comment (gavinr) [reply src/model.js]: 👍

File: src/model.js
Comment (gavinr) [reply src/model.js:67]: [resolving because this same issue is discussed below]

File: src/model.js
Comment (haoliangyu) [reply src/model.js:118]: Does the query require a geometry/location so it needs a default value?

Comment (gavinr) [reply src/model.js:118]: yes

**Review by haoliangyu — COMMENTED** [review]
The code looks good to me. If you can add the `filtersApplied` properties mentioned by @rgwozdz in the GeoJSON output, I can stamp it approved.

**Comment by gavinr** [comment]
@haoliangyu that is in place, thanks. After merge I will add issues to follow-up on the additional issues left above

**Review by haoliangyu — APPROVED** [review]

File: src/model.js
Comment (rgwozdz) [inline src/model.js:157]: I don't want to hold this up any longer, but if the geometry parameter is invalid, it should return a 400 error to the client.  Just add a TODO for now.

**Review by rgwozdz — CHANGES_REQUESTED** [review]

File: src/model.js
Comment (gavinr) [reply src/model.js:157]: fixed

**Review by rgwozdz — APPROVED** [review]

**Comment by rgwozdz** [comment]
I'll release and publish shortly.

**Pull request merged**
